lib/route/link/ip6gre.c

```diff
@@ -0,0 +1,788 @@
+#define IP6GRE_ATTR_OKEY          (1 << 4)
+#define IP6GRE_ATTR_LOCAL         (1 << 5)
+#define IP6GRE_ATTR_REMOTE        (1 << 6)
+#define IP6GRE_ATTR_TTL           (1 << 7)
+#define IP6GRE_ATTR_ENCAPLIMIT    (1 << 8)
+#define IP6GRE_ATTR_FLOWINFO      (1 << 9)
+#define IP6GRE_ATTR_FLAGS         (1 << 10)
+
+struct ip6gre_info
+{
+	uint8_t             ttl;
+	uint8_t             encaplimit;
+	uint16_t            iflags;
+	uint16_t            oflags;
+	uint32_t            ikey;
+	uint32_t            okey;
+	uint32_t            link;
+	uint32_t            flowinfo;
+	uint32_t            flags;
+	struct in6_addr     local;
+	struct in6_addr     remote;
+	uint32_t            ip6gre_mask;
+};
+
+static  struct nla_policy ip6gre_policy[IFLA_GRE_MAX + 1] = {
+	[IFLA_GRE_LINK]         = { .type = NLA_U32 },
+	[IFLA_GRE_IFLAGS]       = { .type = NLA_U16 },
+	[IFLA_GRE_OFLAGS]       = { .type = NLA_U16 },
+	[IFLA_GRE_IKEY]         = { .type = NLA_U32 },
+	[IFLA_GRE_OKEY]         = { .type = NLA_U32 },
+	[IFLA_GRE_LOCAL]        = { .minlen = sizeof(struct in6_addr) },
+	[IFLA_GRE_REMOTE]       = { .minlen = sizeof(struct in6_addr) },
+	[IFLA_GRE_TTL]          = { .type = NLA_U8 },
+	[IFLA_GRE_ENCAP_LIMIT]  = { .type = NLA_U8 },
+	[IFLA_GRE_FLOWINFO]     = { .type = NLA_U32 },
+	[IFLA_GRE_FLAGS]        = { .type = NLA_U32 },
+};
+
+static int ip6gre_alloc(struct rtnl_link *link)
+{
+	struct ip6gre_info *ip6gre;
+
+	if (link->l_info)
+		memset(link->l_info, 0, sizeof(*ip6gre));
+	else {
+		ip6gre = calloc(1, sizeof(*ip6gre));
+		if (!ip6gre)
+			return -NLE_NOMEM;
+
+		link->l_info = ip6gre;
+	}
+
+	return 0;
+}
+
+static int ip6gre_parse(struct rtnl_link *link, struct nlattr *data,
+		        struct nlattr *xstats)
+{
+	struct nlattr *tb[IFLA_GRE_MAX + 1];
+	struct ip6gre_info *ip6gre;
+	int err;
+
+	NL_DBG(3, "Parsing IP6GRE link info\n");
+
+	err = nla_parse_nested(tb, IFLA_GRE_MAX, data, ip6gre_policy);
+	if (err < 0)
+		goto errout;
+
+	err = ip6gre_alloc(link);
+	if (err < 0)
+		goto errout;
+
+	ip6gre = link->l_info;
+
+	if (tb[IFLA_GRE_LINK]) {
+		ip6gre->link = nla_get_u32(tb[IFLA_GRE_LINK]);
+		ip6gre->ip6gre_mask |= IP6GRE_ATTR_LINK;
+	}
+
+	if (tb[IFLA_GRE_IFLAGS]) {
+		ip6gre->iflags = nla_get_u16(tb[IFLA_GRE_IFLAGS]);
+		ip6gre->ip6gre_mask |= IP6GRE_ATTR_IFLAGS;
+	}
+
+	if (tb[IFLA_GRE_OFLAGS]) {
+		ip6gre->oflags = nla_get_u16(tb[IFLA_GRE_OFLAGS]);
+		ip6gre->ip6gre_mask |= IP6GRE_ATTR_OFLAGS;
+	}
+
+	if (tb[IFLA_GRE_IKEY]) {
+		ip6gre->ikey = nla_get_u32(tb[IFLA_GRE_IKEY]);
+		ip6gre->ip6gre_mask |= IP6GRE_ATTR_IKEY;
+	}
+
+	if (tb[IFLA_GRE_OKEY]) {
+		ip6gre->okey = nla_get_u32(tb[IFLA_GRE_OKEY]);
+		ip6gre->ip6gre_mask |= IP6GRE_ATTR_OKEY;
+	}
+
+	if (tb[IFLA_GRE_LOCAL]) {
+		nla_memcpy(&ip6gre->local, tb[IFLA_GRE_LOCAL], sizeof(struct in6_addr));
+		ip6gre->ip6gre_mask |= IP6GRE_ATTR_LOCAL;
+	}
+
+	if (tb[IFLA_GRE_REMOTE]) {
+		nla_memcpy(&ip6gre->remote, tb[IFLA_GRE_LOCAL], sizeof(struct in6_addr));
```

Comment:
IFLA_GRE_LOCAL should be IFLA_GRE_REMOTE on this line

Reply:
please feel invitated to fix the branch and just force-push to github. That will update the pull-request.